Python/aot_ceval_jit.c

```diff
@@ -2246,6 +2309,133 @@ void* jit_func(PyCodeObject* co, PyThreadState* tstate) {
                 deferred_vs_pop1_owned(Dst, arg4_idx);
             }
             deferred_vs_apply(Dst);
+
+            char wrote_inline_cache = 0;
+            if (opcode == CALL_METHOD) {
+                CallMethodHint* hint = Dst->call_method_hints;
+                Dst->call_method_hints = hint->next;
+
+                if (hint->attr && hint->meth_found) {
+                    int num_args = oparg + hint->meth_found; // number of arguments to the function, including a potential "self"
+                    int num_vs_args = num_args + 1; // number of python values; one more than the number of arguments since it includes the callable
+
+                    if (hint->attr->ob_type == &PyMethodDescr_Type) {
+                        PyMethodDescrObject* method = (PyMethodDescrObject*)hint->attr;
+                        void* funcptr = method->d_method->ml_meth;
+
+                        if (funcptr && _PyObject_RealIsSubclass((PyObject*)hint->type, (PyObject *)PyDescr_TYPE(method))) {
+                            wrote_inline_cache = 1;
+
+                            // Strategy:
+                            // First guard on tstate->use_tracing == 0
+                            // It looks like we have to guard on this variable, even though we already
+                            // guarded on ceval->tracing_possible, because it looks like a profile
+                            // function will set use_tracing but not tracing_possible
+                            //
+                            // Then guard that meth != NULL.
+                            // We need this to verify that the object in the "self" stack slot
+                            // is actually the self object and not a non-method attribute.
+                            //
+                            // Then guard that self->ob_type == hint->type
+                            //
+                            // We skip the recursion check since we know we did one when
+                            // entering this python frame.
+
+                            JIT_ASSERT(sizeof(tstate->use_tracing) == 4, "");
+                            | mov eax, [tstate + offsetof(PyThreadState, use_tracing)]
+                            | test eax, eax
+                            | jne >1
+
+                            | mov arg1, [vsp - 8 * num_vs_args] // callable
+                            | test arg1, arg1
+                            | je >1
+
```

Comment:
Replacing this mov+test compinations with the following code should save two instructions. I guess this will likely not change perf but I think it makes the code also a bit more readable because it's clear that it's just a guard and later code is not reading the register.
```
                            | cmp dword [tstate + offsetof(PyThreadState, use_tracing)], 0
                            | jne >1

                            | cmp dword [vsp - 8 * num_vs_args], 0 // callable
                            | je >1
```

Reply:
good think you also catched my `cmp dword error` which should have been a `qword` like you wrote.